src/transformers/cache_utils.py

```diff
@@ -192,246 +222,541 @@ def get_mask_sizes(self, cache_position: torch.Tensor, layer_idx: int) -> tuple[
         The masks are then prepared according to the given lengths (kv_length, kv_offset) and patterns (i.e. sliding_window, chunk_size),
         for each layer.
         """
-        query_length = cache_position.shape[0]
-        past_seen_tokens = self.get_seq_length()
-        kv_length = query_length + past_seen_tokens
-        return kv_length, 0
+        kv_length, kv_offset = self.layers[layer_idx].get_mask_sizes(
+            cache_position, self.key_cache[layer_idx], self.value_cache[layer_idx]
+        )
+        return kv_length, kv_offset
 
+    ### Wrappers for layer operations and properties ###
 
-@dataclass
-class CacheConfig:
+    def get_max_cache_shape(self, layer_idx: int = 0) -> int:
+        """Returns the maximum sequence length of the cache object. DynamicLayer does not have a maximum length."""
+        return self.layers[layer_idx].get_max_cache_shape(self.key_cache[layer_idx], self.value_cache[layer_idx])
```

Comment:
Etc etc everywhere, you get the idea 😁
```suggestion
    def get_max_cache_shape(self, layer_idx: int = 0) -> int:
        """Returns the maximum sequence length of the cache object. DynamicLayer does not have a maximum length."""
        return self.layers[layer_idx].get_max_cache_shape()
```